File: packages/core/src/analytics/projects.ts
/**
 * Project analytics — pure aggregation over SessionHead[], the project-level
 * counterpart to dashboard.ts. No HTTP, no DB: the caller supplies the groups
 * and the sessions, this module owns the arithmetic.
 *
 * Metrics are inclusive of sub-sessions and `sessionCount` is recomputed from
 * the sessions actually handed in, so a windowed caller gets windowed counts
 * instead of the cache-wide count carried by ProjectGroup.
 */
import type { ProjectGroup, SessionHead } from "../types/index.js";
import type {
  ApiProjectAgentStat,
  ApiProjectGroup,
  ApiProjectSummary,
  SessionTree,
  SessionTreeNode,
} from "../contract/index.js";
import { getProjectIdentityKey } from "../contract/project-identity.js";
import {
  buildSessionTree,
  filterSessionTreeEntriesByActivityWindow,
} from "../contract/session-tree.js";
import { getSessionAgentName } from "./dashboard.js";
import type { DashboardCostFacts } from "./cost-facts.js";
import { visitAttributedCosts, visitAttributedUsage } from "./cost-attribution.js";

interface ProjectMetrics {
  sessions: number;
  messages: number;
  tokens: number;
  cost: number;
  hasEstimatedCost: boolean;
  agentStats: Map<string, ApiProjectAgentStat>;
}

export function summarizeProjects(projects: readonly ApiProjectGroup[]): ApiProjectSummary {
  const summary: ApiProjectSummary = {
    projects: projects.length,
    sessions: 0,
    tokens: 0,
    cost: 0,
    latestActivity: null,
  };

  for (const project of projects) {
    summary.sessions += project.sessionCount;
    summary.tokens += project.tokens;
    summary.cost += project.cost;
    if (project.lastActivity != null) {
      summary.latestActivity =
        summary.latestActivity == null
          ? project.lastActivity
          : Math.max(summary.latestActivity, project.lastActivity);
    }
  }

  return summary;
}

function emptyMetrics(): ProjectMetrics {
  return {
    sessions: 0,
    messages: 0,
    tokens: 0,
    cost: 0,
    hasEstimatedCost: false,
    agentStats: new Map(),
  };
}

/**
 * Fold per-session totals into the project groups they belong to. Sessions with
 * no resolved project identity contribute to nothing.
 */
export function attachProjectMetrics(
  projects: ProjectGroup[],
  sessions: SessionHead[],
): ApiProjectGroup[] {
  const tree = buildSessionTree(sessions);
  return attachProjectMetricsToEntries(projects, tree, tree.entries);
}

export function attachProjectMetricsFromTree(
  projects: ProjectGroup[],
  tree: SessionTree,
  from?: number,
  to?: number,
  costFacts?: DashboardCostFacts | null,
): ApiProjectGroup[] {
  return attachProjectMetricsToEntries(
    projects,
    tree,
    filterSessionTreeEntriesByActivityWindow(tree, from, to),
    from,
    to,
    costFacts,
  );
}

function attachProjectMetricsToEntries(
  projects: ProjectGroup[],
  tree: SessionTree,
  entries: SessionTreeNode[],
  from?: number,
  to?: number,
  costFacts?: DashboardCostFacts | null,
): ApiProjectGroup[] {
  const metrics = new Map<string, ProjectMetrics>();

  for (const node of entries) {
    const identity = node.session.project_identity;
    if (!identity) continue;

    const key = getProjectIdentityKey(identity);
    let current = metrics.get(key);
    if (!current) {
      current = emptyMetrics();
      metrics.set(key, current);
    }

    current.sessions += 1;

    const agentName = getSessionAgentName(node.session);
    const agent = current.agentStats.get(agentName);
    if (agent) {
      agent.sessions += 1;
    } else {
      current.agentStats.set(agentName, {
        name: agentName,
        sessions: 1,
        messages: 0,
        tokens: 0,
        cost: 0,
      });
    }
  }

  visitAttributedUsage(
    tree,
    { from, to: to ?? Number.POSITIVE_INFINITY, facts: costFacts },
    ({ entry, messages, totalTokens }) => {
      const identity = entry.session.project_identity;
      if (!identity) return;
      const key = getProjectIdentityKey(identity);
      let current = metrics.get(key);
      if (!current) {
        current = emptyMetrics();
        metrics.set(key, current);
      }
      current.messages += messages;
      current.tokens += totalTokens;

      const agentName = getSessionAgentName(entry.session);
      const agent = current.agentStats.get(agentName);
      if (agent) {
        agent.messages += messages;
        agent.tokens += totalTokens;
      } else {
        current.agentStats.set(agentName, {
          name: agentName,
          sessions: 0,
          messages,
          tokens: totalTokens,
          cost: 0,
        });
      }
    },
  );

  visitAttributedCosts(
    tree,
    { from, to: to ?? Number.POSITIVE_INFINITY, facts: costFacts },
    ({ entry, cost, source }) => {
      const identity = entry.session.project_identity;
      if (!identity) return;
      const key = getProjectIdentityKey(identity);
      let current = metrics.get(key);
      if (!current) {
        current = emptyMetrics();
        metrics.set(key, current);
      }
      current.cost += cost;
      if (source === "estimated") current.hasEstimatedCost = true;

      const agentName = getSessionAgentName(entry.session);
      const agent = current.agentStats.get(agentName);
      if (agent) agent.cost += cost;
      else {
        current.agentStats.set(agentName, {
          name: agentName,
          sessions: 0,
          messages: 0,
          tokens: 0,
          cost,
        });
      }
    },
  );

  return projects.map((project) => {
    const metric = metrics.get(
      getProjectIdentityKey({ kind: project.identityKind, key: project.identityKey }),
    );
    return {
      ...project,
      sessionCount: metric?.sessions ?? 0,
      messages: metric?.messages ?? 0,
      tokens: metric?.tokens ?? 0,
      cost: metric?.cost ?? 0,
      cost_source:
        metric && metric.cost > 0
          ? metric.hasEstimatedCost
            ? "estimated"
            : "recorded"
          : undefined,
      agentStats: [...(metric?.agentStats.values() ?? [])].sort((a, b) => b.sessions - a.sessions),
    };
  });
}

File: crates/codesesh-core/src/analytics/projects.rs
use super::{DashboardCostFacts, DashboardScope, Metrics, attribution, project_key};
use crate::{
    contract::{CostSource, SessionHead},
    query::{SessionTree, in_window},
};
use serde_json::{Value, json};
use std::collections::HashMap;

#[derive(Default)]
struct ProjectMetrics {
    total: Metrics,
    estimated: bool,
    agents: Vec<(String, Metrics)>,
}
impl ProjectMetrics {
    fn agent(&mut self, name: &str) -> &mut Metrics {
        let index = self
            .agents
            .iter()
            .position(|(n, _)| n == name)
            .unwrap_or_else(|| {
                self.agents.push((name.to_owned(), Metrics::default()));
                self.agents.len() - 1
            });
        &mut self.agents[index].1
    }
}

pub fn attach_project_metrics(
    projects: &[Value],
    sessions: &[SessionHead],
    from: Option<f64>,
    to: Option<f64>,
    facts: Option<&DashboardCostFacts>,
) -> Vec<Value> {
    let tree = SessionTree::new(sessions);
    let mut metrics = HashMap::<String, ProjectMetrics>::new();
    for &entry in &tree.entries {
        let s = tree.sessions[entry];
        if !in_window(s.time_updated, from, to) {
            continue;
        }
        let m = metrics.entry(project_key(s)).or_default();
        m.total.sessions += 1;
        m.agent(&s.reference.agent_name.to_lowercase()).sessions += 1;
    }
    let contributions = attribution::contributions(
        &tree,
        &DashboardScope::default(),
        from,
        to.unwrap_or(f64::INFINITY),
        facts,
    );
    for c in &contributions {
        if !c.has_usage {
            continue;
        }
        let s = tree.sessions[c.entry];
        let m = metrics.entry(project_key(s)).or_default();
        m.total.messages += c.usage.messages;
        m.total.tokens += c.usage.tokens;
        let a = m.agent(&s.reference.agent_name.to_lowercase());
        a.messages += c.usage.messages;
        a.tokens += c.usage.tokens;
    }
    for c in &contributions {
        if c.cost <= 0.0 {
            continue;
        }
        let s = tree.sessions[c.entry];
        let m = metrics.entry(project_key(s)).or_default();
        m.total.cost += c.cost;
        if c.source == CostSource::Estimated {
            m.estimated = true;
        }
        m.agent(&s.reference.agent_name.to_lowercase()).cost += c.cost;
    }
    projects
        .iter()
        .map(|p| {
            let mut p = p.clone();
            let key = format!(
                "{}:{}",
                p["identityKind"].as_str().unwrap_or(""),
                p["identityKey"].as_str().unwrap_or("")
            );
            let mut empty = ProjectMetrics::default();
            let m = metrics.get_mut(&key).unwrap_or(&mut empty);
            p["sessionCount"] = json!(m.total.sessions);
            p["messages"] = json!(m.total.messages);
            p["tokens"] = json!(m.total.tokens);
            p["cost"] = json!(m.total.cost);
            if m.total.cost > 0.0 {
                p["cost_source"] = json!(if m.estimated { "estimated" } else { "recorded" });
            } else if let Some(p) = p.as_object_mut() {
                p.remove("cost_source");
            }
            m.agents.sort_by(|a, b| b.1.sessions.cmp(&a.1.sessions));
            p["agentStats"] = json!(
                m.agents
                    .iter()
                    .map(|(name, m)| {
                        let mut a = m.value();
                        a["name"] = json!(name);
                        a
                    })
                    .collect::<Vec<_>>()
            );
            p
        })
        .collect()
}

pub fn summarize_projects(projects: &[Value]) -> Value {
    let mut sessions = 0_u64;
    let mut tokens = 0.0;
    let mut cost = 0.0;
    let mut latest = None;
    for project in projects {
        sessions += project["sessionCount"].as_u64().unwrap_or(0);
        tokens += project["tokens"].as_f64().unwrap_or(0.0);
        cost += project["cost"].as_f64().unwrap_or(0.0);
        if let Some(time) = project["lastActivity"].as_f64() {
            latest = Some(latest.map_or(time, |old: f64| old.max(time)));
        }
    }
    json!({"projects":projects.len(),"sessions":sessions,"tokens":tokens,"cost":cost,"latestActivity":latest})
}
